shared/ui/textField/TextField.tsx

```diff
@@ -1,102 +1,123 @@
 
 import Close from '@/assets/icons/close.svg?url';
 import Eye from '@/assets/icons/eye-outline.svg?url';
 import SearchOutline from '@/assets/icons/searchOutline.svg?url';
 import { Typography } from '@/shared/ui/typography';
+import { clsx } from 'clsx';
 import Image from 'next/image';
 
 import s from 'shared/ui/textField/TextField.module.scss';
 
-// type TextFieldType = 'input' | 'textarea';
+type SharedProps = {
+  className?: string;
+  disabled?: boolean;
+  error?: string;
+  label?: string;
+  onBlur?: () => void;
+  onChange: (value: string) => void;
+  onFocus?: () => void;
+  placeholder?: string;
+  required?: boolean;
+  value: string;
+};
 
+// If you pass this type as a generic parameter, TS will prevent you from using props specific to union members.
+// To be able to pass these props, you can extract the desired union member using the Extract utility type.
+// https://www.typescriptlang.org/docs/handbook/utility-types.html#extracttype-union
 export type TextFieldProps =
-  // <T extends TextFieldType = 'input'>
+  | (SharedProps & {
+      as?: 'input';
+      inputType?: 'password' | 'search' | 'text';
+    })
+  | (SharedProps & { as: 'textarea' });
 
-  {
-    as?: 'input' | 'textarea';
-    className?: string;
-    disabled?: boolean;
-    errors?: string;
-    inputtype?: 'password' | 'text';
-    isSearchInput?: boolean;
-    label?: string;
-    onChange: (e: string) => void;
-    onFocus?: () => void;
-    placeholder?: string;
-    required?: boolean;
-    value: string;
-  };
+// Forces the 'ref' type match the 'as' prop type. You should not specify the 'as' props dynamically.
+// Use only the literal value to avoid possible errors with non-existent properties of the target element pointed by the 'ref'.
+// prettier-ignore
+export type TextFieldComponent = {
+  (props: Extract<TextFieldProps, { as: 'textarea'}> & { ref?: ForwardedRef<HTMLTextAreaElement> }): ReactNode;
+  (props: Extract<TextFieldProps, { as?: 'input'}> & { ref?: ForwardedRef<HTMLInputElement> }): ReactNode;
+}
 
-export const TextField = forwardRef((props: TextFieldProps, ref) => {
-  const {
-    as,
-    className,
-    disabled,
-    errors,
-    inputtype = 'text',
-    isSearchInput,
-    label,
-    onChange,
-    placeholder = 'email',
-    required,
-    value,
-  } = props;
+export const TextField: TextFieldComponent = forwardRef<
+  HTMLInputElement | HTMLTextAreaElement,
+  TextFieldProps
+>(({ className, error, label, onChange, required, ...props }, ref) => {
+  const [showPassword, setShowPassword] = useState(false);
 
-  const [type, setType] = useState(inputtype);
+  const handleChange = (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) =>
+    onChange(e.currentTarget.value);
 
-  const changeInputType = () => {
-    type === 'password' ? setType('text') : setType('password');
-  };
+  const sharedClassName = clsx(s.input, error && s.error, className);
```

Comment:
Here and elsewhere in the code, the `className` calculations performed by ternary are replaced by a call to `clsx`.